src/baldr/incident_singleton.h

```diff
@@ -97,24 +65,66 @@ struct incident_singleton_t {
     if (state->signal.wait_until(lock, when) == std::cv_status::timeout) {
       throw std::runtime_error("Unable to initialize incident watcher in the configured time period");
     }
-    // let this run forever
-    watcher.detach();
+  }
+
+  /**
+   * The default interrupt for the watcher threads main loop. Since its not set, the watcher loop
+   * will run forever. This method is provided so the unit tests can exercise the constructor
+   * @return
+   */
+  virtual std::function<bool(size_t)> interrupt() {
+    return {};
+  }
+
+  /**
+   * Read the contents of a file into an incident tile
+   * @param filename   name of the file on the file system to read into memory
+   * @return a shared pointer with the data of the tile or an empty pointer if it could not be read
+   */
+  static std::shared_ptr<const valhalla::IncidentsTile> read_tile(const std::string& filename) {
+    // open the file for reading. its normal for this to fail when the file has been removed
+    std::ifstream file(filename, std::ios::in | std::ios::binary);
+    if (!file.is_open()) {
+      return {};
+    }
+
+    // prepare a stream for parsing
+    std::string buffer((std::istreambuf_iterator<char>(file)), std::istreambuf_iterator<char>());
+    google::protobuf::io::ArrayInputStream as(static_cast<const void*>(buffer.c_str()),
+                                              buffer.size());
+    google::protobuf::io::CodedInputStream cs(
+        static_cast<google::protobuf::io::ZeroCopyInputStream*>(&as));
+
+    // try to parse the stream
+    std::shared_ptr<valhalla::IncidentsTile> tile(new valhalla::IncidentsTile);
+    if (!tile->ParseFromCodedStream(&cs)) {
+      LOG_WARN("Incident Watcher failed to parse " + filename);
+      return {};
+    }
+
+    // dont store empty tiles no point
+    if (tile->locations_size() == 0) {
+      return {};
+    }
+
+    // hand back something that isnt modifyable
+    return std::const_pointer_cast<const valhalla::IncidentsTile>(tile);
   }
 
   /**
    * Updates the tile in the states cache
    * @param state     the state to update
    * @param tile_id   the tile id we are loading
    * @param path      the path to the tile file
+   * @param hint      a pointer to an existing iterator into the cache
```

Comment:
this was added for performance in the detecting of removed tiles, explained below